lib/shelley/bench/restore-bench.hs

```diff
@@ -664,25 +728,32 @@ bench_restoration proxy tr wlTr socket np vData benchname wallets traceToDisk ta
 
                 let (wid0, wname0, _) = head wallets
                 results <- benchmarks proxy w wid0 wname0 benchname restorationTime
-                Aeson.encodeFile resultsFilepath results
+                saveBenchmarkPoints benchname results
                 forM_ wallets $ \(wid, _, _) ->
                     unsafeRunExceptT (W.deleteWallet w wid)
                 pure $ SomeBenchmarkResults results
   where
     fst' (x,_,_) = x
-    timelogFilepath = T.unpack benchname <> ".timelog"
-    resultsFilepath = T.unpack benchname <> ".json"
-
-    withWalletLayerTracer act
-        | traceToDisk =
-            withFile timelogFilepath WriteMode $ \h -> do
-                -- Use a custom tracer to output (time, blockHeight) to a file
-                -- each time we apply blocks.
-                let fileTr = Tracer $ \msg -> do
-                        liftIO . B8.hPut h . T.encodeUtf8 . (<> "\n") $ msg
-                        hFlush h
-                act $ traceProgressForPlotting fileTr
-        | otherwise   = act nullTracer
+
+walletWorkerLogToBlocHeaders :: WalletWorkerLog -> [BlockHeader]
```

Comment:
Using `NonEmpty BlockHeader` here would save a couple of `toList` and also obsolete the case `[]` in `traceBlockHeadersProgressForPlotting`.

The type `NonEmpty` denotes a nonempty list, it is provided `Data.List.NonEmpty` and somewhat useful.

Reply:
I couldn't solve it because failing the pattern matching cannot be resolved in something better then a mempty